File: src/app/land.service.ts
import { Injectable, signal, computed } from '@angular/core';
import { Land, LandStatus, LandType, LandFilter } from './land.model';

@Injectable({
  providedIn: 'root'
})
export class LandService {
  private landsSignal = signal<Land[]>(this.getInitialLands());
  
  // Public read-only signals
  lands = this.landsSignal.asReadonly();
  
  // Computed signal for available lands
  availableLands = computed(() => 
    this.landsSignal().filter(land => land.status === LandStatus.AVAILABLE)
  );

  constructor() {
    this.loadFromLocalStorage();
  }

  private getInitialLands(): Land[] {
    return [
      {
        id: '1',
        title: 'Prime Residential Land in Abdoun',
        description: 'Beautiful residential plot in the heart of Abdoun with stunning views. Perfect for building your dream home.',
        location: 'Abdoun, Amman',
        area: 1000,
        price: 500000,
        status: LandStatus.AVAILABLE,
        type: LandType.RESIDENTIAL,
        features: ['Mountain View', 'Close to Schools', 'Quiet Neighborhood', 'Utilities Available'],
        images: ['🏡'],
        ownerName: 'Ahmad Mansour',
        ownerContact: '+962-79-1234567',
        datePosted: new Date('2026-01-15'),
        coordinates: { lat: 31.9567, lng: 35.8782 }
      },
      {
        id: '2',
        title: 'Commercial Land - City Center',
        description: 'Strategic commercial location perfect for retail or office development. High foot traffic area.',
        location: 'Downtown, Amman',
        area: 2500,
        price: 1200000,
        status: LandStatus.AVAILABLE,
        type: LandType.COMMERCIAL,
        features: ['Main Street', 'High Traffic', 'Public Transport', 'Parking Available'],
        images: ['🏢'],
        ownerName: 'Sara Al-Khalil',
        ownerContact: '+962-77-9876543',
        datePosted: new Date('2026-01-20'),
        coordinates: { lat: 31.9539, lng: 35.9106 }
      },
      {
        id: '3',
        title: 'Agricultural Land - Jordan Valley',
        description: 'Fertile agricultural land with water access. Ideal for farming or agricultural projects.',
        location: 'Jordan Valley',
        area: 10000,
        price: 300000,
        status: LandStatus.AVAILABLE,
        type: LandType.AGRICULTURAL,
        features: ['Water Source', 'Fertile Soil', 'Irrigation System', 'Road Access'],
        images: ['🌾'],
        ownerName: 'Mohammed Rashid',
        ownerContact: '+962-78-5555555',
        datePosted: new Date('2026-01-10'),
        coordinates: { lat: 32.0523, lng: 35.6157 }
      },
      {
        id: '4',
        title: 'Industrial Plot - Sahab',
        description: 'Large industrial plot suitable for factories or warehouses. Easy highway access.',
        location: 'Sahab, Amman',
        area: 5000,
        price: 800000,
        status: LandStatus.PENDING,
        type: LandType.INDUSTRIAL,
        features: ['Highway Access', 'Industrial Zone', 'Three-Phase Power', 'Large Area'],
        images: ['🏭'],
        ownerName: 'Khalid Omar',
        ownerContact: '+962-79-7777777',
        datePosted: new Date('2026-01-05'),
        coordinates: { lat: 31.8699, lng: 36.0015 }
      },
      {
        id: '5',
        title: 'Mixed-Use Development Land',
        description: 'Perfect for mixed-use development project. Residential and commercial zoning approved.',
        location: 'Sweifieh, Amman',
        area: 3000,
        price: 950000,
        status: LandStatus.AVAILABLE,
        type: LandType.MIXED_USE,
        features: ['Mixed Zoning', 'Modern Area', 'Shopping Nearby', 'Investment Opportunity'],
        images: ['🏘️'],
        ownerName: 'Layla Haddad',
        ownerContact: '+962-77-3333333',
        datePosted: new Date('2026-01-25'),
        coordinates: { lat: 31.9342, lng: 35.8641 }
      }
    ];
  }

  private loadFromLocalStorage(): void {
    const stored = localStorage.getItem('lands');
    if (stored) {
      try {
        const parsed = JSON.parse(stored);
        // Convert date strings back to Date objects
        const lands = parsed.map((land: any) => ({
          ...land,
          datePosted: new Date(land.datePosted)
        }));
        this.landsSignal.set(lands);
      } catch (error) {
        console.error('Error loading lands from localStorage:', error);
      }
    }
  }

  private saveToLocalStorage(): void {
    localStorage.setItem('lands', JSON.stringify(this.landsSignal()));
  }

  getAllLands(): Land[] {
    return this.landsSignal();
  }

  getLandById(id: string): Land | undefined {
    return this.landsSignal().find(land => land.id === id);
  }

  addLand(land: Omit<Land, 'id' | 'datePosted'>): Land {
    const newLand: Land = {
      ...land,
      id: this.generateId(),
      datePosted: new Date()
    };
    
    this.landsSignal.update(lands => [...lands, newLand]);
    this.saveToLocalStorage();
    return newLand;
  }

  updateLand(id: string, updates: Partial<Land>): boolean {
    const index = this.landsSignal().findIndex(land => land.id === id);
    if (index === -1) return false;

    this.landsSignal.update(lands => {
      const newLands = [...lands];
      newLands[index] = { ...newLands[index], ...updates };
      return newLands;
    });
    
    this.saveToLocalStorage();
    return true;
  }

  deleteLand(id: string): boolean {
    const initialLength = this.landsSignal().length;
    this.landsSignal.update(lands => lands.filter(land => land.id !== id));
    
    if (this.landsSignal().length < initialLength) {
      this.saveToLocalStorage();
      return true;
    }
    return false;
  }

  filterLands(filter: LandFilter): Land[] {
    return this.landsSignal().filter(land => {
      if (filter.searchTerm) {
        const term = filter.searchTerm.toLowerCase();
        const matchesSearch = 
          land.title.toLowerCase().includes(term) ||
          land.description.toLowerCase().includes(term) ||
          land.location.toLowerCase().includes(term);
        if (!matchesSearch) return false;
      }

      if (filter.status && land.status !== filter.status) return false;
      if (filter.type && land.type !== filter.type) return false;
      if (filter.location && !land.location.toLowerCase().includes(filter.location.toLowerCase())) return false;
      if (filter.minPrice !== undefined && land.price < filter.minPrice) return false;
      if (filter.maxPrice !== undefined && land.price > filter.maxPrice) return false;
      if (filter.minArea !== undefined && land.area < filter.minArea) return false;
      if (filter.maxArea !== undefined && land.area > filter.maxArea) return false;

      return true;
    });
  }

  private generateId(): string {
    return Date.now().toString(36) + Math.random().toString(36).substr(2);
  }

  formatPrice(price: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0
    }).format(price);
  }

  formatArea(area: number): string {
    return `${area.toLocaleString()} m²`;
  }

  getStatusLabel(status: LandStatus): string {
    const labels = {
      [LandStatus.AVAILABLE]: 'Available',
      [LandStatus.PENDING]: 'Pending',
      [LandStatus.SOLD]: 'Sold'
    };
    return labels[status];
  }

  getTypeLabel(type: LandType): string {
    const labels = {
      [LandType.RESIDENTIAL]: 'Residential',
      [LandType.COMMERCIAL]: 'Commercial',
      [LandType.AGRICULTURAL]: 'Agricultural',
      [LandType.INDUSTRIAL]: 'Industrial',
      [LandType.MIXED_USE]: 'Mixed Use'
    };
    return labels[type];
  }
}

File: Services/LandService.cs
using Land.Marketplace.Models;

namespace Land.Marketplace.Services;

public interface ILandService
{
    IEnumerable<LandItem> GetAllLands();
    LandItem? GetLandById(string id);
    LandItem AddLand(CreateLandRequest request);
    bool UpdateLand(string id, UpdateLandRequest request);
    bool DeleteLand(string id);
    IEnumerable<LandItem> FilterLands(LandFilter filter);
}

public class LandService : ILandService
{
    private readonly List<LandItem> _lands;

    public LandService()
    {
        _lands = GetInitialLands();
    }

    private List<LandItem> GetInitialLands()
    {
        return new List<LandItem>
        {
            new LandItem
            {
                Id = "1",
                Title = "Prime Residential Land in Abdoun",
                Description = "Beautiful residential plot in the heart of Abdoun with stunning views. Perfect for building your dream home.",
                Location = "Abdoun, Amman",
                Area = 1000,
                Price = 500000,
                Status = LandStatus.AVAILABLE,
                Type = LandType.RESIDENTIAL,
                Features = new List<string> { "Mountain View", "Close to Schools", "Quiet Neighborhood", "Utilities Available" },
                Images = new List<string> { "🏡" },
                OwnerName = "Ahmad Mansour",
                OwnerContact = "+962-79-1234567",
                DatePosted = new DateTime(2026, 1, 15),
                Coordinates = new Coordinates { Lat = 31.9567, Lng = 35.8782 }
            },
            new LandItem
            {
                Id = "2",
                Title = "Commercial Land - City Center",
                Description = "Strategic commercial location perfect for retail or office development. High foot traffic area.",
                Location = "Downtown, Amman",
                Area = 2500,
                Price = 1200000,
                Status = LandStatus.AVAILABLE,
                Type = LandType.COMMERCIAL,
                Features = new List<string> { "Main Street", "High Traffic", "Public Transport", "Parking Available" },
                Images = new List<string> { "🏢" },
                OwnerName = "Sara Al-Khalil",
                OwnerContact = "+962-77-9876543",
                DatePosted = new DateTime(2026, 1, 20),
                Coordinates = new Coordinates { Lat = 31.9539, Lng = 35.9106 }
            },
            new LandItem
            {
                Id = "3",
                Title = "Agricultural Land - Jordan Valley",
                Description = "Fertile agricultural land with water access. Ideal for farming or agricultural projects.",
                Location = "Jordan Valley",
                Area = 10000,
                Price = 300000,
                Status = LandStatus.AVAILABLE,
                Type = LandType.AGRICULTURAL,
                Features = new List<string> { "Water Source", "Fertile Soil", "Irrigation System", "Road Access" },
                Images = new List<string> { "🌾" },
                OwnerName = "Mohammed Rashid",
                OwnerContact = "+962-78-5555555",
                DatePosted = new DateTime(2026, 1, 10),
                Coordinates = new Coordinates { Lat = 32.0523, Lng = 35.6157 }
            },
            new LandItem
            {
                Id = "4",
                Title = "Industrial Plot - Sahab",
                Description = "Large industrial plot suitable for factories or warehouses. Easy highway access.",
                Location = "Sahab, Amman",
                Area = 5000,
                Price = 800000,
                Status = LandStatus.PENDING,
                Type = LandType.INDUSTRIAL,
                Features = new List<string> { "Highway Access", "Industrial Zone", "Three-Phase Power", "Large Area" },
                Images = new List<string> { "🏭" },
                OwnerName = "Khalid Omar",
                OwnerContact = "+962-79-7777777",
                DatePosted = new DateTime(2026, 1, 5),
                Coordinates = new Coordinates { Lat = 31.8699, Lng = 36.0015 }
            },
            new LandItem
            {
                Id = "5",
                Title = "Mixed-Use Development Land",
                Description = "Perfect for mixed-use development project. Residential and commercial zoning approved.",
                Location = "Sweifieh, Amman",
                Area = 3000,
                Price = 950000,
                Status = LandStatus.AVAILABLE,
                Type = LandType.MIXED_USE,
                Features = new List<string> { "Mixed Zoning", "Modern Area", "Shopping Nearby", "Investment Opportunity" },
                Images = new List<string> { "🏘️" },
                OwnerName = "Layla Haddad",
                OwnerContact = "+962-77-3333333",
                DatePosted = new DateTime(2026, 1, 25),
                Coordinates = new Coordinates { Lat = 31.9342, Lng = 35.8641 }
            }
        };
    }

    public IEnumerable<LandItem> GetAllLands()
    {
        return _lands;
    }

    public LandItem? GetLandById(string id)
    {
        return _lands.FirstOrDefault(l => l.Id == id);
    }

    public LandItem AddLand(CreateLandRequest request)
    {
        var newLand = new LandItem
        {
            Id = GenerateId(),
            Title = request.Title,
            Description = request.Description,
            Location = request.Location,
            Area = request.Area,
            Price = request.Price,
            Status = request.Status,
            Type = request.Type,
            Features = request.Features,
            Images = request.Images,
            OwnerName = request.OwnerName,
            OwnerContact = request.OwnerContact,
            DatePosted = DateTime.UtcNow,
            Coordinates = request.Coordinates
        };

        _lands.Add(newLand);
        return newLand;
    }

    public bool UpdateLand(string id, UpdateLandRequest request)
    {
        var land = _lands.FirstOrDefault(l => l.Id == id);
        if (land == null) return false;

        if (request.Title != null) land.Title = request.Title;
        if (request.Description != null) land.Description = request.Description;
        if (request.Location != null) land.Location = request.Location;
        if (request.Area.HasValue) land.Area = request.Area.Value;
        if (request.Price.HasValue) land.Price = request.Price.Value;
        if (request.Status.HasValue) land.Status = request.Status.Value;
        if (request.Type.HasValue) land.Type = request.Type.Value;
        if (request.Features != null) land.Features = request.Features;
        if (request.Images != null) land.Images = request.Images;
        if (request.OwnerName != null) land.OwnerName = request.OwnerName;
        if (request.OwnerContact != null) land.OwnerContact = request.OwnerContact;
        if (request.Coordinates != null) land.Coordinates = request.Coordinates;

        return true;
    }

    public bool DeleteLand(string id)
    {
        var land = _lands.FirstOrDefault(l => l.Id == id);
        if (land == null) return false;

        _lands.Remove(land);
        return true;
    }

    public IEnumerable<LandItem> FilterLands(LandFilter filter)
    {
        var query = _lands.AsEnumerable();

        if (!string.IsNullOrEmpty(filter.SearchTerm))
        {
            var term = filter.SearchTerm.ToLower();
            query = query.Where(l =>
                l.Title.ToLower().Contains(term) ||
                l.Description.ToLower().Contains(term) ||
                l.Location.ToLower().Contains(term));
        }

        if (filter.Status.HasValue)
            query = query.Where(l => l.Status == filter.Status.Value);

        if (filter.Type.HasValue)
            query = query.Where(l => l.Type == filter.Type.Value);

        if (!string.IsNullOrEmpty(filter.Location))
            query = query.Where(l => l.Location.ToLower().Contains(filter.Location.ToLower()));

        if (filter.MinPrice.HasValue)
            query = query.Where(l => l.Price >= filter.MinPrice.Value);

        if (filter.MaxPrice.HasValue)
            query = query.Where(l => l.Price <= filter.MaxPrice.Value);

        if (filter.MinArea.HasValue)
            query = query.Where(l => l.Area >= filter.MinArea.Value);

        if (filter.MaxArea.HasValue)
            query = query.Where(l => l.Area <= filter.MaxArea.Value);

        return query.ToList();
    }

    private string GenerateId()
    {
        return Guid.NewGuid().ToString("N")[..12];
    }
}
